vendor/github.com/ethereum/go-ethereum/whisper/notifications/delivery.go

```diff
@@ -0,0 +1,136 @@
+package notifications
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/message"
+	whisper "github.com/ethereum/go-ethereum/whisper/whisperv5"
+)
+
+// DeliveryState holds references to a message delivered through the network.
+type DeliveryState struct {
+	IsP2P bool
+	P2P   *whisper.P2PMessageState
+	RPC   *whisper.RPCMessageState
+}
+
+// DeliverySubscriber defines a function type for subscrubers.
+type DeliverySubscriber func(DeliveryState)
+
+// P2PDeliverySubscriber defines a function type for receiving p2p message status.
+type P2PDeliverySubscriber func(*whisper.P2PMessageState)
+
+// RPCDeliverySubscriber defines a function type for receiving rpc message status.
+type RPCDeliverySubscriber func(*whisper.RPCMessageState)
+
+// DeliveryService defines a notification implementation for listening to message status
+// events.
+type DeliveryService struct {
+	sml  sync.RWMutex
+	subs []DeliverySubscriber
+}
+
+// SendP2PState delivers a rpc message status to all subscribers.
+func (d *DeliveryService) SendP2PState(state whisper.P2PMessageState) {
+	if state.Timestamp.IsZero() {
+		state.Timestamp = time.Now()
+	}
+	d.sendState(DeliveryState{P2P: &state})
+}
+
+// SendRPCState delivers a rpc message status to all subscribers.
+func (d *DeliveryService) SendRPCState(state whisper.RPCMessageState) {
+	if state.Timestamp.IsZero() {
+		state.Timestamp = time.Now()
+	}
+	d.sendState(DeliveryState{RPC: &state})
+}
+
+// SendState delivers envelope with status to all subscribers.
+func (d *DeliveryService) sendState(mds DeliveryState) {
+	d.sml.RLock()
+	defer d.sml.RUnlock()
+
+	for _, item := range d.subs {
```

Comment:
Maybe `for _, subscribe := d.subs` {?

Reply:
`subscribe` is a verb. Singular `sub` or `subscription` would IMHO be fine.